core/silkworm/crypto/ecdsa.cpp

```diff
@@ -86,4 +94,34 @@ std::optional<Bytes> recover(ByteView message, ByteView signature, bool odd_y_pa
     return out;
 }
 
+std::optional<evmc::address> public_key_to_address(const Bytes& public_key) {
+    assert(public_key.length() == 65);
+    if (public_key[0] == 4u) {
+        // Ignore first byte of public key
+        const auto key_hash{ethash::keccak256(public_key.data() + 1, public_key.length() - 1)};
+        evmc::address recovered_address{};
+        std::memcpy(&recovered_address.bytes[0], &key_hash.bytes[kHashLength - kAddressLength], kAddressLength);
+        return recovered_address;
+    }
+    return std::nullopt;
+}
+
+std::optional<evmc::address> recover_address(ByteView message, ByteView signature, bool odd_y_parity) {
+    const auto recovered_public_key{recover(message, signature, odd_y_parity)};
+    if (!recovered_public_key.has_value()) {
+        return std::nullopt;
+    }
+    return public_key_to_address(recovered_public_key.value());
+}
+
+std::optional<evmc::address> recover_address(secp256k1_context* context, ByteView message, ByteView signature,
```

Comment:
Perhaps make `context` the last argument and default it to `nullptr` in order to avoid duplicated versions of `recover_address`? (and `recover` would call `create_context` if the passed one is `nullptr`)

Reply:
Yeah right.

Reply:
Hmm wait ... need a bit of refactor then ... that would break the static context instantiated in `recover`

Reply:
> Hmm wait ... need a bit of refactor then ... that would break the static context instantiated in `recover`

Well, we can have something like
```
std::optional<Bytes> recover(ByteView message, ByteView signature, bool odd_y_parity,
                                                  secp256k1_context* context) {
    static secp256k1_context* static_context{create_context()};
    if (context == nullptr) {
         context = static_context;
    }
    ...
}
```

Reply:
Commit 3f60bd8